src/EVM/Fuzz.hs

```diff
@@ -0,0 +1,238 @@
+                            }
+
+findVarProp :: Prop -> State CollectVars Prop
+findVarProp p = mapPropM go p
+  where
+    go :: forall a. Expr a -> State CollectVars (Expr a)
+    go = \case
+      e@(Var a) -> do
+        s :: CollectVars <- get
+        put CollectVars {vars=Set.insert (Var a) s.vars, vals = s.vals}
+        pure e
+      e@(Lit a) -> do
+        s :: CollectVars <- get
+        put $ s{vals=Set.insert a s.vals}
+        pure e
+      e -> pure e
+
+
+extractVars :: [Prop] -> CollectVars
+extractVars ps = execState (mapM_ findVarProp ps) initVarsState
+
+
+--- Buf extraction
+newtype CollectBufs = CollectBufs { bufs :: Set.Set (Expr Buf) }
+  deriving (Show)
+
+initBufsState :: CollectBufs
+initBufsState = CollectBufs { bufs = Set.empty }
+
+extractBufs :: [Prop] -> [Expr Buf]
+extractBufs ps = Set.toList bufs
+ where
+  CollectBufs bufs = execState (mapM_ findBufProp ps) initBufsState
+
+findBufProp :: Prop -> State CollectBufs Prop
+findBufProp p = mapPropM go p
+  where
+    go :: forall a. Expr a -> State CollectBufs (Expr a)
+    go = \case
+      e@(AbstractBuf a) -> do
+        s <- get
+        put $ s{bufs=Set.insert (AbstractBuf a) s.bufs}
+        pure e
+      e -> pure e
+
+--- Store extraction
+data CollectStorage = CollectStorage { addrs :: Set.Set (Expr EAddr)
+                                     , keys :: Set.Set W256
+                                     , values :: Set.Set W256
+                                     }
+  deriving (Show)
+
+initStorageState :: CollectStorage
+initStorageState = CollectStorage { addrs = Set.empty, keys = Set.empty, values = Set.fromList [0x0, 0x1, Expr.maxLit] }
+
+extractStorage :: [Prop] -> CollectStorage
+extractStorage ps = execState (mapM_ findStorageProp ps) initStorageState
+
+findStorageProp :: Prop -> State CollectStorage Prop
+findStorageProp p = mapPropM go p
+  where
+    go :: forall a. Expr a -> State CollectStorage (Expr a)
+    go = \case
+      e@(AbstractStore a) -> do
+        s <- get
+        put $  CollectStorage {addrs=Set.insert a s.addrs, keys=s.keys, values=s.values}
+        pure e
+      e@(SLoad (Lit val) _) -> do
+        s <- get
+        put $ s{keys=Set.insert val s.keys}
+        pure e
+      e@(SStore _ (Lit val) _) -> do
+          s <- get
+          put $ s{values=Set.insert val s.values}
+          pure e
+      e -> pure e
+
+-- Var value and TX value generation
+getVals :: CollectVars -> Gen (Map (Expr EWord) W256)
+getVals vars = do
+    bufs <- go (Set.toList vars.vars) mempty
+    addTxStuff bufs
+  where
+    addTxStuff :: Map (Expr EWord) W256 -> Gen (Map (Expr EWord) W256)
+    addTxStuff a = do
+      val <- frequency [ (20, pure 0)
+                       , (1, getRndW256)
+                       ]
+      pure $ Map.insert TxValue val a
+    go :: [Expr EWord] -> Map (Expr EWord) W256 -> Gen (Map (Expr EWord) W256)
+    go [] valMap = pure valMap
+    go (a:ax) valMap = do
+      pickKnown :: Bool <- arbitrary
```

Comment:
I understand now that this is where the known literals are used. But why is it helpful to pick a value from the set of known literals?

Reply:
It allows us to hop over some tricky ITE-s, like `if (a == 57)` where it'd be hard to figure out 57 (especially out of 2**256). At least that's the idea :) This technique is used in most of fuzzers, and all fuzzers I know allow you to add such magic numbers if you know them in advance for your program. For example, if some parts of the parser may never start if it doesn't see the magic constant `0xfea3fadc`, then you can add that magic constant to the fuzzer and hopefully it'll manage to fuzz more parts of your code.

I check for this in this test:

```
    , test "fuzz-simple-fixed-value3" $ do
      Just c <- solcRuntime "MyContract"
        [i|
        contract MyContract {
          function func(uint a, uint b) public {
            assert(((a != 1337) && (b != 99)));
          }
        }
        |]
```

I hope this helps :)

Reply:
oh that's really nice! Thanks for explaining!

Reply:
:)